src/reader/lexer.rs

```diff
@@ -272,33 +275,48 @@ impl PullLexer {
         }
 
         // Reading characters
-        let mut buffer = Vec::with_capacity(4);
+        self.buffer.clear();
         loop {
+            // Read a byte in order to read an utf-8 code point
             if let Some(byte) = b.bytes().next().and_then(|i| i.ok()) {
-                buffer.push(byte);
+                self.buffer.push(byte);
             } else {
+                // Nothing to read left in reader.
                 break;
             }
 
-            let s = match String::from_utf8(buffer) {
+            // Try to get one unicode code point.
+            // As we added only a byte, we can get at most a utf-8 string with
+            //  a single code point.
+            let cp = match str::from_utf8(&self.buffer) {
                 Ok(s) => {
-                    buffer = Vec::with_capacity(4);
-                    s
+                    match s.chars().next() {
+                        Some(chr) => {
+                            chr
+                        },
+                        None => {
+                            // Should never get there, because the string
+                            // contains exactly one code point.
+                            continue;
```

Comment:
I think that if this branch is never supposed to be reached, it should be stubbed with `unreachable!()`.